File: swift-extensions/mixpanel/MixpanelAnalyticsService.swift
import Foundation
import Mixpanel
import TRIKOT_FRAMEWORK_NAME

public class MixpanelAnalyticsService: AnalyticsService {
    public init(enableAnalytics: Bool = true) {
        enabled = enableAnalytics
    }

    public var enabled: Bool {
        didSet {
            if enabled {
                Mixpanel.mainInstance().optInTracking()
            }
            else {
                Mixpanel.mainInstance().optOutTracking()
            }
        }
    }

    public func identifyUser(userId: String, properties: [String: Any]) {
        Mixpanel.mainInstance().identify(distinctId: userId)
        Mixpanel.mainInstance().people.set(properties: properties.asMixpanelProperties)
    }

    public func incrementUserProperties(incrementalProperties: [String: Any]) {
        Mixpanel.mainInstance().people.increment(properties: incrementalProperties.asMixpanelProperties)
    }

    public func logout() {
        Mixpanel.mainInstance().reset()
    }

    public func setSuperProperties(properties: [String: Any]) {
        Mixpanel.mainInstance().registerSuperProperties(properties.asMixpanelProperties)
    }

    public func setUserProperties(properties: [String: Any]) {
        Mixpanel.mainInstance().people.set(properties: properties.asMixpanelProperties)
    }

    public func trackEvent(event: AnalyticsEvent, properties: [String: Any]) {
        Mixpanel.mainInstance().track(event: event.name, properties: properties.asMixpanelProperties)
    }

    public func unsetAllSuperProperties() {
        Mixpanel.mainInstance().clearSuperProperties()
    }

    public func unsetSuperProperties(propertyNames: [String]) {
        propertyNames.forEach { property in
            Mixpanel.mainInstance().unregisterSuperProperty(property)
        }
    }

    public var name = "MixpanelAnalytics"
}

// This is needed to convert Kotlin type to Swift types for Mixpanel to uses these values
extension Dictionary where Key == String {
    var asMixpanelProperties: Properties {
        return compactMapValues { value in
            if let intValue = value as? Int {
                return intValue
            }
            else if let doubleValue = value as? Double {
                return doubleValue
            }
            else if let stringValue = value as? String {
                return stringValue
            }
            else {
                return "\(value)"
            }
        }
    }
}

File: mixpanel-ktx/src/main/java/com/mirego/trikot/analytics/MixpanelAnalyticsService.kt
package com.mirego.trikot.analytics

import android.content.Context
import com.mixpanel.android.mpmetrics.MixpanelAPI
import org.json.JSONObject

class MixpanelAnalyticsService(
    context: Context,
    mixpanelToken: String,
    analyticsEnabled: Boolean = true
) : AnalyticsService {
    private val mixpanelAnalytics =
        MixpanelAPI.getInstance(context, mixpanelToken, !analyticsEnabled)

    override var isEnabled: Boolean
        get() = !mixpanelAnalytics.hasOptedOutTracking()
        set(value) {
            if (value) {
                mixpanelAnalytics.optInTracking()
            } else {
                mixpanelAnalytics.optOutTracking()
            }
        }

    override val name: String = "MixpanelAnalytics"

    override fun identifyUser(userId: String, properties: AnalyticsPropertiesType) {
        mixpanelAnalytics.identify(userId)
        mixpanelAnalytics.people.identify(userId)
        mixpanelAnalytics.people.set(properties.asJSONProperties())

        mixpanelAnalytics
    }

    override fun incrementUserProperties(incrementalProperties: AnalyticsIncrementalProperties) {
        mixpanelAnalytics.people.increment(incrementalProperties)
    }

    override fun logout() {
        mixpanelAnalytics.reset()
    }

    override fun setSuperProperties(properties: AnalyticsPropertiesType) {
        mixpanelAnalytics.registerSuperProperties(properties.asJSONProperties())
    }

    override fun setUserProperties(properties: AnalyticsPropertiesType) {
        mixpanelAnalytics.people.set(properties.asJSONProperties())
    }

    override fun unsetAllSuperProperties() {
        mixpanelAnalytics.clearSuperProperties()
    }

    override fun trackEvent(event: AnalyticsEvent, properties: AnalyticsPropertiesType) {
        mixpanelAnalytics.track(event.name, properties.asJSONProperties())
    }

    override fun unsetSuperProperties(propertyNames: List<String>) {
        propertyNames.forEach { name ->
            mixpanelAnalytics.unregisterSuperProperty(name)
        }
    }

    private fun AnalyticsPropertiesType.asJSONProperties(): JSONObject {

        return JSONObject().also { props ->
            forEach { (type, value) ->
                props.put(type, value)
            }
        }
    }
}
